### PR TITLE
Make the mobile orientation mode configurable

### PR DESCRIPTION
**Pull request by miterosan**
Only for android.

Can be tested using TestSceneOrientation.

### CODE REVIEW

File: osu.Framework/Platform/IWindow.cs [hunk 1]
```diff
@@ -87,5 +87,7 @@ public interface IWindow : IGameWindow
         /// Available resolutions for full-screen display.
         /// </summary>
         IEnumerable<DisplayResolution> AvailableResolutions { get; }
+
+        IEnumerable<WindowOrientationMode> SupportedOrientationModes { get; }
```
Comment (swoolcock) [inline osu.Framework/Platform/IWindow.cs]: Add xmldoc.

File: osu.Framework/Configuration/WindowOrientationMode.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+{
+    public enum WindowOrientationMode
+    {
+        Landscape,
+        UpsideDownLandscape,
+        SensorLandscape,
+        Portrait,
+        UpsideDownPortrait,
+        SensorPortait,
+        Sensor
```
Comment (swoolcock) [inline osu.Framework/Configuration/WindowOrientationMode.cs]: What's the difference between the various `Sensor` modes?

File: osu.Framework.Android/AndroidGameWindow.cs [hunk 3]
```diff
@@ -26,19 +28,61 @@ public override bool Focused
...
+
+            config.BindWith(FrameworkSetting.WindowOrientationMode, orientationMode);
         }
 
         protected override IEnumerable<WindowMode> DefaultSupportedWindowModes => new WindowMode[]
         {
             Configuration.WindowMode.Fullscreen,
         };
 
+        public override IEnumerable<WindowOrientationMode> SupportedOrientationModes => new[]
```
Comment (swoolcock) [inline osu.Framework.Android/AndroidGameWindow.cs]: Be aware that framework consumers will be able to specify supported orientation modes from their startup projects (both Android and iOS), so this is not guaranteed to be correct.

**Comment by peppy** [comment]
why is this user configurable? i can't think of a single scenario or application that offers this functionality to users (the mobile operating system usually provides this functionality).

it should be exposed to the framework consumer, not end user.

File: osu.Framework/Configuration/WindowOrientationMode.cs (hunk 2 above)
Comment (miterosan) [reply osu.Framework/Configuration/WindowOrientationMode.cs]: The sensor modes detect if you put your phone upside down and rotate the view accordingly.

**Comment by miterosan** [comment]
I agree that the framework consumer should decide the orientation or allow the users to configure it.

Most mobile games go with landscape_sensor or portrait, because they design their games around that screen orientation.

This PR is not required if [this suggested change] (https://github.com/ppy/osu/pull/4021#discussion_r291842588) gets accepted. 
Setting it to FullSensor allows users to rotate the screen how they want while osu is never on the left, the right or upside-down.

> the mobile operating system usually provides this functionality

Sadly this is only partially true on android.
On most android versions you can only completely disable sensor based orientation change, or have it fully on. There is no way to specify landscape or portrait as the preferred orientation.
When you have it disabled and go to your homescreen (always portrait) and then into an app that is on fullsensor/none, it will just keep the orientation.

@peppy how should be continue here?

**Comment by swoolcock** [comment]
> On most android versions you can only completely disable sensor based orientation change, or have it fully on. There is no way to specify landscape or portrait as the preferred orientation.

This is the same on iOS for iPhone/iPod Touch.  There's auto-rotate (if the app supports it), or portrait-lock.  iPad does not have this limitation, you can lock it to whichever is the current orientation.  Having said that, games can easily rotate their rendering manually anyway.

I agree that this is not something that should be user-configurable at a framework level, as the user should choose orientation using their device's settings, in whichever way is natively supported by the operating system.

**Comment by peppy** [comment]
It's fine to have this at framework, but it cannot be in config. It should be a GameWindow or GameHow virtual property.

**Comment by miterosan** [comment]
not doing this.

**Pull request closed without merging**